File: services/raffle/handlers/create.js
const awsSDK = require('aws-sdk');
const { generateShortCode } = require('../utils');

const documentClient = new awsSDK.DynamoDB.DocumentClient();
const RAFFLE_TABLE = process.env.RAFFLE_TABLE;

module.exports.handler = (event, context, callback) => {
  const body = JSON.parse(event.body);
  const name = body.name;
  const admins = body.admins ? body.admins : [];
  const timestamp = new Date().getTime();

  const shortcode = generateShortCode();
  const params = {
    TableName: RAFFLE_TABLE,
    Item: {
      shortcode,
      name,
      timestamp,
      admins: documentClient.createSet(admins),
    },
    Expected: {
      shortcode: {
        // We don't want to overwrite an existing shortcode
        Exists: false,
      },
    },
  };

  documentClient.put(params, function(error, data) {
    if (error) {
      console.log(error);
      const response = {
        statusCode: 503,
        body: JSON.stringify({
          message: 'Could not create raffle. Please try again',
        }),
      };
      callback(null, response);
    } else {
      const response = {
        statusCode: 201,
        body: JSON.stringify({
          shortcode,
          name,
        }),
      }
      callback(null, response);
    }
  });
};

File: services/raffle/handlers/create.py
import json
import logging

from lib import db, auth

logger = logging.getLogger()
logger.setLevel(logging.INFO)


def handler(event, context):

    try:
        email= auth.get_email(event['headers'])
    except auth.MissingAuthentication:
        return {
            "statusCode": 401,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"error": "Please authorize with an Authorization header."})
        }
    except auth.InvalidAuthentication as e:
        return {
            "statusCode": 401,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"error": e.message })
        }

    if not auth.can_create_raffle(email):
        return {
            "statusCode": 403,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"error": "User not authorized to to create a raffle."})
        }

    body = json.loads(event.get('body'))
    name = body.get('name')
    description = body.get('description')

    if not name:
        return {
            "statusCode": 400,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"error": "Request body must include a 'name' key"})
        }
    if not description:
        return {
            "statusCode": 400,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"error": "Request body must include a 'description' key"})
        }

    admins = body.get('admins', [])

    try:
        shortcode = db.create_raffle(name, description, admins)
    except Exception as e:
        logger.exception(e)
        return {
            "statusCode": 503,
            "headers": {
                "Access-Control-Allow-Origin" : "*",
                "Access-Control-Allow-Credentials" : True
            },
            "body": json.dumps({"message": "Could not create raffle. Please try again."})
        }

    response = {
        "statusCode": 200,
        "headers": {
            "Access-Control-Allow-Origin" : "*",
            "Access-Control-Allow-Credentials" : True
        },
        "body": json.dumps({
            'shortcode': shortcode,
            'name': name,
            'description': description
        })
    }

    return response
